### ISSUE TITLE
ledger-mode setup for hledger needs documenting

### ISSUE THREAD

**Issue by simonmichael** [simonmichael/hledger#367]

**Comment by taksuyu** [comment]
Since I'm working on getting this working, I don't mind writing something up once I got it working reasonably well.

**Comment by simonmichael** [comment]
FYI:
```elisp
; Simon's ledger-mode tweaks

(custom-set-variables

 ; neutralise some ledgerisms
 '(ledger-binary-path (expand-file-name "~/.local/bin/hledger"))
 '(ledger-mode-should-check-version nil)
 '(ledger-init-file-name " ")

 ; move default amount position right allowing longer account names
 '(ledger-post-amount-alignment-column 64)

 ; disable distracting highlight
 '(ledger-highlight-xact-under-point nil)

 )

; automatically show new transactions from hledger add or hledger-web
(add-hook 'ledger-mode-hook 'auto-revert-tail-mode)

; M-1 to collapse all transactions to one line, M-0 to reset. Useful for quickly scanning
(global-set-key "\M-0" (lambda () (interactive) (set-selective-display (* tab-width 0))))
(global-set-key "\M-1" (lambda () (interactive) (set-selective-display (* tab-width 1))))
(add-hook 'ledger-mode-hook (lambda () (setq tab-width 4)))

; enable some highlighting for CSV rules files
(add-to-list 'auto-mode-alist '("\\.rules$" . conf-mode))

; useful when running reports in a shell buffer
(defun highlight-negative-amounts nil (interactive)
  (highlight-regexp "\\(\\$-\\|-\\$\\)[.,0-9]+" (quote hi-red-b))
)
```

**Comment by simonmichael** [comment]
```elisp
; recognise more file suffixes
(add-to-list 'auto-mode-alist '("\\.\\(h?ledger\\|journal\\|j\\)$" . ledger-mode))

; enable orgstruct minor mode, TAB while on a * comment expands/collapses org node
(add-hook 'ledger-mode-hook 'orgstruct-mode)
```

**Comment by simonmichael** [comment]
https://github.com/narendraj9/dotfiles/tree/master/.emacs.d/packages/rest/hledger-mode should be investigated too, I haven't tried it.

**Comment by simonmichael** [comment]
Other editor modes like https://github.com/ledger/vim-ledger would also be nice to know about.

**Comment by taksuyu** [comment]
I seem to be comfortable using this configuration. I imagine there is more to look into and fixing little things.

```lisp
(setq ledger-binary-path "hledger")
(setq ledger-mode-should-check-version nil)
(add-to-list 'auto-mode-alist '("\\.\\(h?ledger\\|journal\\|j\\)$" . ledger-mode))

(defvar ledger-report-balance
  (list "bal" (concat ledger-binary-path " -f %(ledger-file) bal")))

(defvar ledger-report-reg
  (list "reg" (concat ledger-binary-path " -f %(ledger-file) reg")))

(defvar ledger-report-payee
  (list "payee" (concat ledger-binary-path " -f %(ledger-file) reg @%(payee)")))

(defvar ledger-report-account
  (list "account" (concat ledger-binary-path " -f %(ledger-file) reg %(account)")))

(setq ledger-reports
      (list ledger-report-balance
            ledger-report-reg
            ledger-report-payee
            ledger-report-account))
```

Having hledger specific reports may also benefit.

**Comment by arademaker** [comment]
The C-c C-a with hledger gives me `hledger: command xact is not recognized, run with no command to see a list` what did I miss?

**Comment by simonmichael** [comment]
ledger-mode's C-c C-a runs ledger-add-transaction. On my setup this is somehow working, I get a Transaction: prompt in minibuffer with today's date filled in, and enter inserts the date in the journal. Not terribly useful.

On yours, I guess it assumes the `ledger-binary-path` program has a `xact` command, and expects that to produce a journal entry on stdout. xact is an alias for Ledger's `entry` command.

hledger doesn't have the entry/xact command, though related functionality exists in several other commands (print --match, register-match, find, add). You could possibly make a good enough version with something like this in an executable hledger-xact.sh script in $PATH:
```
#!/bin/sh
hledger print --match $@
```

**Comment by simonmichael** [comment]
PS ledger-mode 2017-09-02 24b43e34 here.

**Comment by arademaker** [comment]
Thanks , but I guess I will also need to redefine the command ledger-add-transaction or create a new one bound to C-c C-a right ?

**Comment by arademaker** [comment]
Hi @simonmichael , it took me a while to try your suggestion but it didn't work as I expected. Can you please help me? Using Ledger (prompt) I can have a new transaction with a new value given other fields automatically filled. For instance, using the `xact` I give a date, description (payee in my case) a new amount and the `account 2`. Like a magic, Ledger deduce the `account 1` for me. 

```$ ledger -f lixo.dat xact 2018/08/10 Recanto R$\ 100 real:cc
2018/08/10 Recanto
    Expenses:Educacao:Escola:Sofia         R$ 100,00
    Assets:AR:Real:CC```

If I give more info, it will work also. In the following case, `housing` is used to find the account `Expense:Housing` and `real:cc` to find `Assets:AR:Real:CC`

```$ ledger -f lixo.dat xact 2018/08/10 Recanto housing R$\ 100 real:cc
2018/08/10 Recanto
    Expenses:Housing                       R$ 100,00
    Assets:AR:Real:CC
```

I haven't figured out how to have similar behaviour with hledger!

**Comment by simonmichael** [comment]
Hi @arademaker, let's handle that question on the mail list and/or IRC channel. Perhaps you can repost there, and add all data needed for others to reproduce it and understand.

**Comment by levitanong** [comment]
When using ledger-report, I'm running into hledger's complaints about unsupported options `--columns` and `--color`.
Ledger mode is passing those options here:
https://github.com/ledger/ledger-mode/blob/d6a67d177bada0a5913019893dfc6015590bcb76/ledger-report.el#L438

So in order for me to use hledger's reports, I've had to:
```
(setq ledger-report-auto-width nil
     ledger-report-use-native-highlighting nil)
```

**Comment by simonmichael** [comment]
-U ?

**Comment by levitanong** [comment]
I'm aware of -U, but I don't know where to set it in emacs. 😅 is there something like `ledger-reconcile-cli-command`?

**Comment by simonmichael** [comment]
It doesn't look like it. I think you'd have to submit a patch to ledger-mode changing `--uncleared` to `-U` (in `ledger-do-reconcile`).

I was thinking -U was a spelling that both tools would recognise. But, now I remember [hledger's -U/--unmarked is different from ledger's -U/--uncleared](http://hledger.org/manual.html#status). The exact match for `ledger --uncleared` is `hledger --unmarked --pending` (`hledger -UP`).

**Comment by simonmichael** [comment]
PS I would be open to a PR making `hledger --uncleared` as a compatibility synonym for `--unmarked --pending`.

**Comment by levitanong** [comment]
Alas, I have no experience in haskell, so I won't be much help here. :(

**Comment by hackeryarn** [comment]
I don't think we are going to reach compatibility on this. The apis seem to have drifted too far apart.

As I attempted to fix the `--uncleared` flag, I continued running into other flags that were incompatible. First it was the `--order` flag, then the `--limit` flag.

@simonmichael  There should be a warning that hledger is not compatible with `ledger-mode`. The only way this could be resolved is either having full backwards compatibility to ledger, or major refactoring to `ledger-mode` to support two different programs.

**Comment by simonmichael** [comment]
Perhaps this was meant for #1002 ?

**Comment by simonmichael** [comment]
Anyway: I'm not sure what problems you are having exactly.. is it that hard to make ledger-mode adapt to hledger here ?

"Not compatible" is too strong - I've used ledger-mode with hledger since the beginning. Some of ledger-mode's features indeed depend on Ledger's specific ui, which we will never match 100%. I mean if someone cared enough they could add a compatibility mode and try to reach 100% ui compatibility. I think it'd be much easier and less costly to generalise ledger-mode a little.

**Comment by simonmichael** [comment]
Getting back to the main issue, perhaps more than this we need to better document ledger-mode + hledger issues. Could someone work on this ? I guess https://hledger.org/Editor-configuration.html is the most likely place.

**Comment by hackeryarn** [comment]
So I just saw that multiple commands needed to be updated, and couldn't just be generalized. e.g., both modes share the `-U` flag but it acts different, the `--sort` flag doesn't exist in hledger.  Also ledger-mode has no hledger specific logic, currently. 

I think the file format editing work perfectly, but most of the command like get balance and reconcile have issues.

I would also be interested in working on adding hledger support to ledger-mode. I can open up an issue and see if they are ok with adding new logic to handle hledger.

**Comment by simonmichael** [comment]
Cool. Probably a small pull request will be a good way.

**Comment by DamienCassou** [comment]
I just sent a [PR to ledger-mode](https://github.com/ledger/ledger-mode/pull/278) to make it possible to get colors from hledger. As a result, it is not required to set `ledger-report-use-native-highlighting` to `nil` anymore as @levitanong [suggested above](https://github.com/simonmichael/hledger/issues/367#issuecomment-433678314). It is required to configure a new variable though:

```emacs
(setq ledger-report-native-highlighting-arguments '("--color=always"))
```

**Comment by DamienCassou** [comment]
I tried to do the same for `ledger-report-auto-width` but failed because of the way ledger-mode passes arguments. I opened an issue there: https://github.com/ledger/ledger-mode/issues/279.

**Comment by simonmichael** [comment]
From srk:
```
(setq ledger-mode-should-check-version nil
      ledger-report-links-in-register nil
      ledger-report-auto-width nil 
      ledger-report-use-native-highlighting nil
      ledger-binary-path "hledger")
```
More notes:

- run/create a report with `[C-u] C-c C-o C-r`
- commands can use expansions, like `%(binary) -f %(ledger-file) bs`  (interpolates `ledger-binary-path` and current buffer's file path)
- in the report buffer: space/backspace to scroll, e to edit the report command, s to save the report command permanently
- the report buffer updates automatically as you save changes to the journal, but the window will disappear unless you make it dedicated (M-x toggle-window-dedicated)
- reports can also be edited via `M-x customize-var ledger-reports`
-  nice thing a ledger-mode + hledger how-to or elisp shim could do is add all the commands & common reports to ledger-reports

**Comment by simonmichael** [comment]
From @DamienCassou:

> my Emacs configuration for hledger: https://github.com/DamienCassou/emacs.d/blob/master/init.el#L677
> my ledger reports are defined here: https://github.com/DamienCassou/emacs.d/blob/master/init.el#L650
> and my flycheck-hledger configuration: https://github.com/DamienCassou/emacs.d/blob/master/init.el#L623

**Comment by simonmichael** [comment]
- A hard-coded use of Ledger's `--date-format` flag was added to ledger-mode's `ledger-exec.el` in April, breaking such hledger compatibility as there was. You'll have to modify ledger-exec.el to get around this. (PR to ledger-mode welcome!)

**Comment by acarrico** [comment]
I've expanded my ledger proxy script given in #1721 to deal with ledger-mode's current C-c C-a ledger-add-transaction behavior. This assumes ISO dates.

ledger.sh:

```
#!/bin/sh

iargs=("$@")
oargs=()
j=0;
date=;
for((i=0; i<${#iargs[@]}; ++i)); do
    case ${iargs[i]} in
        --date-format)
            # drop --date-format and the next arg
            i=$((i+1));
            ;;
        xact)
            # convert "xact" to "print --match"
            oargs[j]=print; oargs[j+1]=--match; j=$((j+2));
            # drop xact argument and stash the date argument
            i=$((i+1));
            date=${iargs[i]};
            ;;
        *)
            # keep any other args:
            oargs[j]=${iargs[i]};
            j=$((j+1));
            ;;
    esac
done

#echo "ledger ${@}" >> ~/tmp/ledger.txt
#echo "hledger ${args[@]}" >> ~/tmp/ledger.txt
#echo "date $date"
#echo "ledger ${@}"
#echo "hledger ${oargs[@]}"

if test "$date"
then
    # substitute the given date for the old date:
    hledger "${oargs[@]}" | sed "1s/....-..-../$date/"
else
    hledger "${oargs[@]}"
fi
```

 Use ```setq ledger-binary-path ledger.sh``` to install your script.

**Comment by DamienCassou** [comment]
Thank you for your script @acarrico, I think it could be very useful to me. I can't make it work though.

For the `sed` to work at the end, the content of `$date` shouldn't contain any slash. But, ledger dates contain slashes by default (see the value of `ledger-default-date-format` which is used by `ledger-format-date` which is the function used to format the date from the minibuffer interactive input). Do you have a specific value for `"input-date-format"` in the variable `ledger-environment-alist`?

The workaround is easy: just change occurrences of "/" in the `sed` argument to something different, e.g., "&".

**Comment by acarrico** [comment]
@DamienCassou, correct, you will have to adjust the script if you don't use ISO standard dates, but maybe you should try them? See: https://github.com/simonmichael/hledger/issues/933

**Comment by Thaodan** [comment]
See also: https://github.com/simonmichael/hledger/issues/621.

**Comment by Thaodan** [comment]
See also: 
https://github.com/ledger/ledger-mode/pull/441

**Comment by Thaodan** [comment]
> ledger-mode's C-c C-a runs ledger-add-transaction. On my setup this is somehow working, I get a Transaction: prompt in minibuffer with today's date filled in, and enter inserts the date in the journal. Not terribly useful.
> 
> On yours, I guess it assumes the `ledger-binary-path` program has a `xact` command, and expects that to produce a journal entry on stdout. xact is an alias for Ledger's `entry` command.
> 
> hledger doesn't have the entry/xact command, though related functionality exists in several other commands (print --match, register-match, find, add). You could possibly make a good enough version with something like this in an executable hledger-xact.sh script in $PATH:
> 
> ```
> #!/bin/sh
> hledger print --match $@
> ```

I managed to bend ledger mode to call  hledger print --match instead.
However is print --match the same thing? From what I see it is to find if there are existing statements of that option.
My patch makes so that hledger is called like this:
`* call-process-region(1 1 "hledger" nil (#<buffer  *temp*> "/tmp/ledger-errorszL2VSn") nil "-f" "-" "print" "--match" "+770€" "Foo" "2025/05/05")`

**Comment by simonmichael** [comment]
I don't remember exactly how `ledger entry` behaves - whatever I said a few years back might be wrong or outdated.

**Comment by simonmichael** [comment]
It would be great if ledger-mode just worked with hledger, or failed smoothly, and if the supported features were well documented somewhere. Thanks for nudging that ledger-mode PR.
